### PR TITLE
Use SingletonSideInput instead of MultiMapSideInput for hashJoin 

### PR DESCRIPTION
**Pull request by daikeshi**
This pr is to fix issue https://github.com/spotify/scio/issues/916

@ravwojdyla @nevillelyh

### CODE REVIEW

**Comment by daikeshi** [comment]
the downside of this fix is that it will take longer to load the side input.

**Comment by ravwojdyla** [comment]
Could you give share some performance numbers on how this will improve performance for at least some synthetic data?

**Comment by daikeshi** [comment]
Yep, I'll run some jobs on test data

**Comment by nevillelyh** [comment]
Take a look at this: https://github.com/spotify/scio/blob/master/scio-test/src/it/scala/com/spotify/ScioBenchmark.scala
You can run it from SBT directly, with & without your commit in PR. You should also add `--regex=.*HashJoin.*` to run that benchmark only.
Probably wanna increase the data size also.

**Comment by nevillelyh** [comment]
@daikeshi did you get a chance to run the benchmark? Would be nice to compare when side input is close/above cache size.

**Comment by ravwojdyla** [comment]
@daikeshi any update here?

**Comment by daikeshi** [comment]
@ravwojdyla @nevillelyh sorry guys, I was on vacation for past two weeks. I'm going to get this done  this week!

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 1]
```diff
@@ -237,9 +237,9 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
    */
   def hashJoin[W: ClassTag](that: SCollection[(K, W)])
   : SCollection[(K, (V, W))] = self.transform { in =>
-    val side = that.asMultiMapSideInput
+    val side = that.map(kv => Map(kv._1 -> Seq(kv._2))).sum.asSingletonSideInput
     in.withSideInputs(side).flatMap[(K, (V, W))] { (kv, s) =>
-      s(side).getOrElse(kv._1, Iterable()).toSeq.map(w => (kv._1, (kv._2, w)))
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Now I think about it, toSeq may be eager and causing major performer penalty.

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 2]
```diff
@@ -237,9 +237,9 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
    */
   def hashJoin[W: ClassTag](that: SCollection[(K, W)])
   : SCollection[(K, (V, W))] = self.transform { in =>
-    val side = that.asMultiMapSideInput
+    val side = that.map(kv => Map(kv._1 -> Seq(kv._2))).sum.asSingletonSideInput
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Even though sum might use Semigroup#sumOption under the hood, I'd make sure if it's comparable with combining into a mutable map.

Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: You can use `asSingletonSideInput(Seq.empty[W])`, and you would need `getOrElse` later on.

**Comment by nevillelyh** [comment]
Per slack discussion, there's no need to keep the multi-map side input based implementation since it breaks the RHS in-memory semantic and is probably just as bad as a full join.

OTOH maybe make `hashLeftJoin` a singleton side input too? Should be trivial.

**Comment by daikeshi** [comment]
baseline asMultiMapSideInput (toSeq): 313 seconds
baseline asMultiMapSideInput (iterator): 269 seconds
singleton side input: 215 seconds

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 3]
```diff
@@ -237,9 +237,9 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
     in.withSideInputs(side).flatMap[(K, (V, W))] { (kv, s) =>
-      s(side).getOrElse(kv._1, Iterable()).iterator.map(w => (kv._1, (kv._2, w)))
+      s(side).getOrElse(kv._1, Seq.empty[W]).map(w => (kv._1, (kv._2, w)))
```
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: using `asSingletonSideInput(Seq.empty[W])` did not work?

Comment (daikeshi) [reply scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: hmm, the side input is a map, `getOrElse` is to return an empty seq for key that doesn't exist in the map.

Comment (ravwojdyla) [reply scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: @daikeshi my bad - you are right.

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 4]
```diff
@@ -237,9 +238,18 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+      combiner.getOrElseUpdate(k, ArrayBuffer.empty[W]) += v
+      combiner
+    } { case (left, right) =>
+        right.foreach { case (k, vs) => left.getOrElseUpdate(k, ArrayBuffer.empty[W]) ++= vs }
+        left
+    }.asSingletonSideInput
+
     in.withSideInputs(side).flatMap[(K, (V, W))] { (kv, s) =>
-      s(side).getOrElse(kv._1, Iterable()).iterator.map(w => (kv._1, (kv._2, w)))
+      s(side).getOrElse(kv._1, ArrayBuffer.empty[W]).map(w => (kv._1, (kv._2, w)))
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Add the `.iterator` back?

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 5]
```diff
@@ -237,9 +238,18 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
    */
   def hashJoin[W: ClassTag](that: SCollection[(K, W)])
   : SCollection[(K, (V, W))] = self.transform { in =>
-    val side = that.asMultiMapSideInput
+    val side = that.combine { case (k, v) =>
+      MMap(k -> ArrayBuffer(v))
```
Comment (daikeshi) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: @nevillelyh @ravwojdyla I updated to use combine with mutable.Map and ArrayBuffer, what do you guys think? looks good?

**Comment by daikeshi** [comment]
Benchmark numbers on hashJoin mutable.Map on ArrayBuffer vs Map on Seq:
```
singleton side input (Map + Seq): 215 seconds
singleton side input (mutable.Map + ArrayBuffer): 189 seconds
```

**Pull request merged**